### ISSUE TITLE
Parcel 2: TypeScript automatic zero configuration not type checking or reading tsconfig.json

### ISSUE THREAD

**Issue by bbugh** [parcel-bundler/parcel#4022]
I've been following #1378 for a long time, and I'm thrilled to see Parcel 2 taking on better first-class support for TypeScript. There's currently some problems with Parcel 2 out of the box, and I didn't see any threads tracking this, so I'm making an issue.

# 🐛 bug report

In 2.0.0-alpha.3.2, given a very basic TypeScript project, there are three issues: 

- the `typescript` package is not automatically installed
- settings in `tsconfig.json` are completely ignored
- typescript errors are silently ignored and compiled anyway

## 🎛 Configuration (.babelrc, package.json, cli command)

Here's all of the files:

- `package.json` created from two commands:
    - `yarn init -y`
    - `yarn install -D parcel@v2`
- `index.html`
	```html
    <!DOCTYPE html>
    <html lang="en">
      <head>
        <meta charset="utf-8" />
        <title>My First Parcel App</title>
        <script src="index.ts"></script>
      </head>
      <body>
        <h1>Hello, World!</h1>
      </body>
    </html>
	```
- `index.ts`
	```typescript
	console.log("Hello, von Nuemann!")
	const failMe: BadType = 5; // this should fail, but it compiles
	```
- optionally, I added a junk `tsconfig.json` to test, and it's obviously being ignored because parcel still compiles:
    ```json
	{
	  ALDIFJ SLDIFJ :LEIRJ LWEIJR;
	}}}}}
      "compilerOptions": {
    ```

## 🤔 Expected Behavior

As part of parcels "zero configuration", parcel automatically detects TypeScript and configures the project to correctly use TypeScript.

- TypeScript compilation or validation should fail by default when there are type errors
- `typescript` package should be installed automatically
- `tsconfig.json` should be read and used by whatever is compiling it, if it exists

## 😯 Current Behavior

- the `typescript` package is not automatically installed
- settings in `tsconfig.json` are completely ignored
- typescript errors are silently ignored and compiled anyway

## 💁 Possible Solution

I'm guessing one of these two potential solutions:

2. The current default `transformer-babel` is correctly configured to behave as expected with TypeScript, extending from #3083
1. `typescript` and `@parcel/transformer-typescript(-tsc?)` is automatically installed when Parcel 2 detects a `.ts` file.

Either one is fine, I think, as long as the TypeScript compilation actually behaves like TypeScript by default!

## 🔦 Context

- Seems to have started in this PR that defaults to `babel` for TypeScript: https://github.com/parcel-bundler/parcel/pull/3083
- Type checking seems to have been merged, but it's not happening: https://github.com/parcel-bundler/parcel/pull/3232

## 💻 Code Sample

See files above

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           | 2.0.0-alpha.3.2 |
| Node             | v10.15.3 |
| npm/Yarn         | 1.21.1 | 
| Operating System | Mac 10.14.6 |

**Comment by DeMoorJasper** [comment]
You have to use a parcelrc file with the transformer-tsc for tsconfig to do anything and for typechecking you need to add the ts validator to the config.

The zero config default we took is just running babel we might however want to detect typescript in the babel transformer so it’s truly zero config

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by bbugh** [comment]
Thank you for the reply. I agree with what you said and understand the current configuration requirements. To clarify, I'm not confused about how it works in Parcel 2, I'm suggesting that current Parcel 2 behavior is incorrect. 

The main value proposition of TypeScript is compile-time typing and checking. Parcel 2 currently ignores all type checking by default. This was a problem with Parcel 1 that [you had previously suggested would be fixed](https://github.com/parcel-bundler/parcel/issues/1378#issuecomment-389785902) in Parcel 2. I am excited to test Parcel 2, and noticed that this incorrect default behavior issue still exists, hence this bug report.

I believe the goal of `parcelrc` is to only be used in situations where Parcel's intelligent defaults do not cover a use case. I suggest that properly supporting TypeScript's type checking should be the default behavior, not behavior that requires extra configuration. 

As an analogy, imagine Parcel 2 ignores sass compilation errors by default, and requires you to configure `parcelrc` to actually show errors. That wouldn't make much sense either!

**Comment by bbugh** [comment]
Thank you for the reply. I agree with what you said and understand the current configuration requirements. To clarify, I'm not confused about how it works in Parcel 2, I'm suggesting that current Parcel 2 behavior is incorrect. 

The main value proposition of TypeScript is compile-time typing and checking. Parcel 2 currently ignores all type checking by default. This was a problem with Parcel 1 that [you had previously suggested would be fixed](https://github.com/parcel-bundler/parcel/issues/1378#issuecomment-389785902) in Parcel 2. I am excited to test Parcel 2, and noticed that this incorrect default behavior issue still exists, hence this bug report.

I believe the goal of `parcelrc` is to only be used in situations where Parcel's intelligent defaults do not cover a use case. I suggest that properly supporting TypeScript's type checking should be the default behavior, not behavior that requires extra configuration. 

As an analogy, imagine Parcel 2 ignores sass compilation errors by default, and requires you to configure `parcelrc` to actually show errors. That wouldn't make much sense either!

**Comment by danielegarciav** [comment]
I also believe type checking should be the default. Using Babel to just get rid of the type annotations and nothing else defeats the main purpose of TypeScript, which is to ensure a strongly typed codebase and every benefit that comes with it.

There are also [some other limitations](https://kulshekhar.github.io/ts-jest/user/babel7-or-ts) of using Babel over TypeScript documented by the team at `ts-jest`, which means that people who rely on TypeScript won't get what they are used to in other projects.

I should add that this would also be the proper way to get error checking on Vue files with Parcel. In the past, it has been recommended to use `tsc --noEmit`, but that command does not work on .vue files.

**Comment by DeMoorJasper** [comment]
Our opinion was that your IDE should catch typing errors when developing and on deploy using a CI/CD service it should be a step before even starting up anything (including Parcel) to speedup the process, similar to how you'd do linting. Typing fails, don't run build/tests/...

**Comment by bbugh** [comment]
I'm quite confused by that position. If the official stance is that external tools should be used for handling compilation errors, why does Parcel report errors for other compilers, like sass? Sass is a superset of CSS with a compiler, much like TypeScript is a superset of JavaScript with a compiler. It seems like you're looking at TypeScript as if it's a linter instead of a separate language with separate compilation and configuration needs.

I propose that TypeScript is treated as a first-class zero-configuration part of Parcel's bundler, using the real Microsoft TypeScript compiler by default, with full type checking in the bundling process. No external tools required, it's just drop in a `ts` file and the appropriate libraries are added and run.

If people want to use babel for whatever reason, that could be configured in `parcelrc`. Not every project that uses TypeScript will want to use Babel, especially because [Babel doesn't support the full TypeScript language spec](https://kulshekhar.github.io/ts-jest/user/babel7-or-ts).

**Comment by DeMoorJasper** [comment]
We have definitely improved TypeScript support greatly since Parcel 1, but we still have a long way to go as you mentioned to become a perfect zero-config tool for TypeScript based projects. (we introduced typechecking, exporting and treeshaking typings, ...)

I've opened a PR to move a step closer to that ideal scenario by enabling typechecking by default, maybe it'll get merged into v2 #4025

**Comment by flybayer** [comment]
I agree with @bbugh.

The current behavior is a deal-breaker for me using Parcel. If I'm using Typescript, I fully expect the build to break if there is a type error.

For contrast, I've been very happy with Next.js's out-of-the-box Typescript support. Just rename a file to .tsx, and you are basically good to go. Next automatically adds a tsconfig.json if it doesn't exist.

I'm not sure if it uses tsc or Babel — either way it's working very well.

**Comment by devongovett** [comment]
> If the official stance is that external tools should be used for handling compilation errors, why does Parcel report errors for other compilers, like sass?

I think there is some confusion between syntactic errors (your code won't parse), and semantic errors (your code has a type error). TypeScript distinguishes between the two. With other tools, Parcel reports only syntactic errors. That doesn't prevent you from having other kinds of errors (e.g. linting, test failures, etc.).

I guess the question is: why would Parcel include TypeScript by default if it isn't going to include ESLint, Jest, etc.? You'd likely want those to fail your build too. So where do we draw the line?

In a real CI setup, you're probably going to run a bunch of tools other than just Parcel: your package manager, your tests, your linting, etc. Is it really so hard to run `tsc` to check types? Why should it be part of bundling?

And during development, you have your IDE to check your types. No need to switch to your terminal to see them. That just sounds inconvenient. So, I'm not sure I really understand why we should do that in Parcel either.

**Comment by danielegarciav** [comment]
> Is it really so hard to run `tsc` to check types?

You can't run `tsc` on .vue files. These files were specifically made to be compiled by a bundler (some standalone tools have been proposed, but none have come to fruition yet), which would then pass the script section to whatever compiler is set.

It is therefore not trivial run type-checking on .vue files using Parcel. There's also this inconsistency: Vue templates are built with the Vue template compiler, Sass files are compiled with the Sass compiler, CoffeeScript files are built with the CoffeeScript compiler. Why is TypeScript not built with the TypeScript compiler?

In reply to this:
> "That doesn't prevent you from having other kinds of errors"

When it comes to making a distinction between type errors and other types of errors, I'd say that types are part of the TypeScript language itself, while linting and testing are not.

**Comment by flybayer** [comment]
I think y'all are viewing it too closely based on the technical implementation instead of from the typical Typescript end-user's perspective.

I view Typescript as totally different than ESLint. I view Typescript as it's own compiled language, like C for example. If a C file has a semantic type error, it won't compile and you can't even run it. **That** is the experience I expect with Typescript.

**Comment by devongovett** [comment]
> Why is TypeScript not built with the TypeScript compiler?

It's faster. We need a Babel AST anyway, so if Babel can compile TS for us, it's faster to do it that way.

That's kinda the wrong question though. The TS compiler has nothing to do with type checking. TS will happily compile files with semantic errors, just like Babel, Sass, or CoffeeScript. In fact, with `transpileModule` (which is what we'd use), TSC does no type checking at all (it cannot, without cross file analysis).

Type checking is a separate task from compiling. That's done by a Validator plugin in Parcel, just like for ESLint.

> If a C file has a semantic type error, it won't compile and you can't even run it

I would find that incredibly annoying during development. I don't write perfect types while I'm still figuring out what to write, and I don't want that to block me trying things. That's why TS separates type checking from compiling: so you can have the best of both worlds.

**Comment by flybayer** [comment]
I can understand how some folks wouldn't want the build to break. 

At the very least, I think Parcel should have a configuration option for causing the build to break.

**Comment by danielegarciav** [comment]
>> If a C file has a semantic type error, it won't compile and you can't even run it

> I would find that incredibly annoying during development.

I think that's where the real opinion split is. I think it's incredibly useful during development. In fact, the sole reason people use TypeScript instead of JavaScript, and indeed the reason it even exists.

> I don't write perfect types while I'm still figuring out what to write, and I don't want that to block me trying things.

It all boils down to discipline. I would say that if you're using TypeScript, you _should_ know what your types are. If you don't know what type, then just use `any`. At least you have documented to other fellow developers that you don't know what type some variable truly is yet.

I would say if you're using TypeScript, you don't just type untyped code temporarily and figure out types as an afterthought. At my workplace we (and I think most programmers around the world) found that "temporary" is the most permanent thing there exists.

If you don't want types, or types are just an afterthought, just use JavaScript. IDEs like VSCode will help you as much as they can even with regular JS files.

All in all, I don't mind having to use config to enable type checking. It is a bit disappointing, yes, but if we are using Babel ASTs and most of the work is done for us, then I think the compromise wouldn't be too bad.

I do have a couple of questions, though:

> That's done by a Validator plugin in Parcel, just like for ESLint.

I'm not fully used to how validators work in Parcel, but would the de facto TypeScript validator able to type-check the script out of a .vue file?

Also, what can we do about [Babel's incompatibility with the full TypeScript spec](https://kulshekhar.github.io/ts-jest/user/babel7-or-ts), including no support for namespaces, const enum, legacy import/export and caret type-casting with JSX enabled?

**Comment by DeMoorJasper** [comment]
> At the very least, I think Parcel should have a configuration option for causing the build to break.

This is possible by adding the typescript validator plugin as we already mentioned. It builds everything and just throws an error at the end as typechecking is totally seperate from compilation like devon mentioned.

I also don't really understand why you'd compare TS with C, TS works completely fine without any types. While you can't even write C without types, typescript is just a superset of JS which makes it basically a fancy linter.

**Comment by mischnic** [comment]
> Also, what can we do about Babel's incompatibility with the full TypeScript spec, including no support for namespaces, const enum, legacy import/export and caret type-casting with JSX enabled?

(Note that Parcel is running TSC via `transpileModule`, which [has limitations as well](https://github.com/Microsoft/TypeScript/issues/5243#issuecomment-147766222))

**Comment by Kinrany** [comment]
I agree that type checking in IDE and CI should be enough, given that there's no type-driven compilation in TS.

I do find it surprising that tsconfig is ignored. I think it's very rare to not have one in a TypeScript project.

Disclaimer: I don't use Parcel (yet). I came from [this Reddit thread](https://www.reddit.com/r/typescript/comments/euu1ml/parcel_2_alpha_typescript_rfc_parcel_2_alphas/).

**Comment by DomiR** [comment]
I agree with @devongovett! It needs to as fast as possible by default (think "blazing fast").

Running full typechecks can still be done when building for production, so you won’t run into any unexpected surprises.

I‘ve been using parcel 1 in every react+typescript project in the past couple of years and it always outperformed my tediously crafted webpack-configs (which is not zero-config at all, me being a webconfig noob). We currently have some >100 file projects and started migrating to CRA for some of Dans "fast refresh" goodness as it was available there first and after dabbling with Flutter and SwiftUI for a bit you will become painfully aware of those build-times... 

Would be interesting to see what people want to configure in tsconfig.json in a zero-config environment (besides strictness rules).

**Comment by devongovett** [comment]
> I'm not fully used to how validators work in Parcel, but would the de facto TypeScript validator able to type-check the script out of a .vue file?

No, that's not possible. TypeScript doesn't know how to import `.vue` files. Parcel cannot magically make it understand them. All the validator plugin does is call the TypeScript language service API to perform type checking. From there, the TS compiler handles your code just like `tsc` on the command line.

**Comment by uglycoyote** [comment]
A couple of thoughts:

Regarding the sentiment that type checking in the IDE should be enough:  My experience is that it is not a thorough check (at least in VSCode).  It only checks the files that you have opened, which in my case is typically a small subset of the project.  There's an [issue](https://github.com/microsoft/vscode/issues/13953#issuecomment-492446129)  requesting them to make the error checking reflect the entire project that has been open since 2016 with 610 thumbs up, but no sign of it being addressed any time soon.   

I tend to agree with @bbugh  that Typescript should be supported properly by default (using tsc), though it seems like a speed vs correctness issue and the trade-off should be configurable.  If babel's transpilation differs in significant ways from tsc, then Typescript users might be in for a surprise when using Parcel, where they encounter these differences.  (e.g. the lack of support for `const enum` is one language feature mentioned in @mischnic's link which Babel apparently does not support.)  I'm not sure how these descrepancies between Babel and Typescript have evolved or will evolve over time.   I would think that for Typescript users the least surprising behaviour for Parcel would be to use tsc, and therefore it's a sensible default.  Users willing to sacrifice correctness to build faster should be able to configure that. 

With regards to @DomiR 's question 

>Would be interesting to see what people want to configure in tsconfig.json in a zero-config environment (besides strictness rules).

The `baseUrl` property comes to mind.  I was a bit shocked and disappointed to find out that Parcel was unable resolve some of my valid `import` statements because it did not obey the `baseUrl` parameter in `tsconfig.json.`.  See my recent issue #4014 and the much older issue #202 regarding this.

In a way this could be viewed as a case where Parcel's behaviour surprised me because it differed from tsc (in terms of how it resolves imports).   But it makes sense that Parcel doesn't resolve files in the same way as tsc if one knows that Parcel doesn't use tsc and that Babel doesn't care about the tsconfig.json either.  But I wasn't really aware that it was cutting corners in this way and therefore I found it to be a surprising and unintuitive experience, and so not a great thing for an out-of-the-box default.

**Comment by devongovett** [comment]
Supporting TypeScript via TSC is pretty common. Hugely popular tools like Next.js, Gatsby, Create React App, and others all do so. Has anyone looked to see if users of those tools have had negative experiences?

**Comment by ianschmitz** [comment]
> Supporting TypeScript via Babel is pretty common. Hugely popular tools like Next.js, Gatsby, Create React App, and others all do so. Has anyone looked to see if users of those tools have had negative experiences?

In CRA we use [fork-ts-checker-webpack-plugin](https://github.com/TypeStrong/fork-ts-checker-webpack-plugin) to run TSC type checking in parallel with babel via @babel/preset-typescript. This has worked fairly well for us and I suspect we would have the same complaints as OP if we removed type checking.

**Comment by devongovett** [comment]
I believe that works exactly how @parcel/validator-typescript works. It also runs in the background to perform type checking, but both still use Babel for actual compilation.

**Comment by moltar** [comment]
Type checking is essential and should not be optional IMO.

**Comment by phaux** [comment]
I like to be able to run type checking, linting, formatting or bundling as a separate command. If I wanted to do everything at once I would just run `tsc && eslint && prettier && parcel` which I usually do in my prepublish script

**Comment by hammypants** [comment]
For me, personally, Parcel + Typescript is unusable, especially under the guise of "zero configuration," with the way it currently works. But I see the other points as well. Better documentation around the expectations of how Parcel + TS is intended to work wouldn't help either-- had to comb many issues to get here.

**Comment by argv-minus-one** [comment]
If you don't run the real TypeScript compiler, you don't support TypeScript, and it's misleading to claim otherwise on your website.

I see that this is due to a performance concern, but that is no excuse. Correctness is always more important than performance. A slow, correct build is merely annoying; a fast, incorrect build is useless.

**Comment by phaux** [comment]
I disagree. I think that a bundler should only do the bundling and the fact that you can transform TS into JS without any typechecking (unlike Elm, Rust, etc) is a big win. If you want to do both, just run `tsc && parcel build` or enable appropriate parcel plugin, but it shouldn't be the default.

**Comment by SamHH** [comment]
There's no point in using TypeScript if you're not going to leverage the type safety it offers you, and you can't meaningfully leverage it without the compiler checking your project as you develop.

Parcel's tagline in this repo is:

> Blazing fast, zero configuration web application bundler

It looks like the heart of the issue is that we have to choose between the two; we can't be "blazing fast" with type-checking, and we surely aren't a "zero-configuration [...] bundler" for TypeScript without type-checking given it's something virtually all TypeScript users will want enabled in tandem with their bundling.

Therefore it clearly needs to be configurable, and all that's left is to determine which of Parcel's goals is most important. In my opinion that's zero-configuration, particularly given that small projects won't notice the performance penalty that type-checking entails.

There is a secondary question that remains regarding Parcel's support being via Babel, which doesn't behave as well as the official TypeScript compiler.

**Comment by SeekDaSky** [comment]
I'd like to weight in and say that I use parcel (incredible work by the way) for the "zero configuration" part. I can setup a project in a matter of minutes and understand what's going on (unlike webpack). I don't care for a few seconds of improvement if the code compiled is wrong.

 I can totally understand your stance, and a choice might be given to the developer of he wants a faster compilation.

But the lack of typecheck by default is really unintuitive for me , I use Typescript mainly for it and I find the faster default to be counterproductive.

**Comment by uglycoyote** [comment]
I find myself agreeing with points from both sides on this issue.  I work exclusively with Typescript and My main reasons for jumping off the Webpack bandwagon were 1) speed and 2) ease of getting up and running.

I was really amazed when I started using Parcel at the response time, when saving my code and seeing things update in the browser in less than a second, where I was used to having it take tens of seconds using Webpack (using a plugin that built Typescript with tsc).   Even though I ideologically fall into the "correctness is more important" camp, in practice I don't know that I would actually want to sacrifice the speed of bundling to ensure that tsc is always run. 

I am sort of bothered though that I was unaware for a long time that it was using babel and not tsc.   I just assumed tsc was the way all typescript was built, and didn't really know that just running it through babel to strip the annotations was even a possibility.  I feel like this should be clearer.  I feel like there should be a warning message with that.    It does say clearly in the [Typescript ](https://parceljs.org/typeScript.html) page that Parcel doesn't do type checking, but maybe I just missed that sentence or didn't fully understand what it meant.

I also feel like, although the goal of Parcel is to be "zero configuration", that should mean that the best defaults are chosen automatically, and not necessarily that it's not configurable. 

I would suggest that if you are building Typescript using babel, the bundler could output `Warning:   Bundling Typescript without using tsc.  To enable type-checking, use "parcelTypescriptBuildMethod: tsc" in your package.json`.  Or, on the other hand if you are using tsc, it could output `Note:  Build speed will be slower because type checking is enabled. For improved performance, use "parcelTypescriptBuildMethod: babel" in your package.json`. (I'm inventing an imaginary config option here)

There's a clear trade off here between speed and checking.  Typescript Correctness people are arguing that it's not a valid Typescript build without the checking.  Fast Bundler people are arguing, that Parcel is a bundler and that's not the bundler's job, so do it out-of-band if you want it.    

Maybe with the "fork-ts-checker" plugin mentioned earlier, one can have their cake and eat it too (fast bundling with type checking on the side), but I haven't tried that.  If that is the best of both worlds for typescript users, maybe that should be the default, zero-configuration setup:  a third mode called `parcelTypescriptBuildMethod: babelWithBackgroundTypecheck` should be the default?

**Comment by mischnic** [comment]
To give an overview over the current implementation:
- Your syntax `parcelTypescriptBuildMethod: babelWithBackgroundTypecheck ` corresponds to
a .parcelrc with
```json
{
  "extends": "@parcel/config-default",
  "validators": {
    "*.{ts,tsx}": ["@parcel/validator-typescript"]
  }
}
```

- `parcelTypescriptBuildMethod: babel` is currently the default
- for `parcelTypescriptBuildMethod: tsc`
```json
{
  "extends": "@parcel/config-default",
  "transformers": {
    "*.{ts,tsx}": ["@parcel/transformer-typescript-tsc"]
  }
}
```

**Comment by uglycoyote** [comment]
@mischnic Thanks, that's very informative, and I think that information, along with a short blurb about the trade-offs involved would make excellent material for a section in the https://parceljs.org/typeScript.html page

**Comment by mischnic** [comment]
> in the parceljs.org/typeScript.html page

(rather: in the page about Typescript in the Parcel 2 docs)

**Comment by argv-minus-one** [comment]
I'm shocked that this is even being seriously discussed. For me, coming from a Java background, compiling without type checking is unthinkably reckless.

**Comment by den-t** [comment]
> `parcelTypescriptBuildMethod: babelWithBackgroundTypecheck ` corresponds to
>   a .parcelrc with
> 
> ```json
> {
>   "extends": "@parcel/config-default",
>   "validators": {
>     "*.{ts,tsx}": ["@parcel/validator-typescript"]
>   }
> }
> ```

Could this be made default?
Any reasons against?

**Comment by ptitjes** [comment]
Hi folks, I tried to implement `~` paths resolution with parcel v2 ala https://parceljs.org/typeScript.html#baseurl-and-paths. But parcel complains it cannot resolve my `~/file.ts` imports. Is this supposed to work in parcel v2 yet ? Should I open an issue ?

Here is what my project tree looks like:
```
- package.json
- tsconfig.json
- .parcelrc
- src/
    - index.html
    - index.ts
    - file.ts
```
Here is my `.parcelrc`:
```json
{
  "extends": "@parcel/config-default",
  "transformers": {
    "*.{ts,tsx}": ["@parcel/transformer-typescript-tsc"],
    "*.svg": ["@parcel/transformer-svgo"]
  },
  "validators": {
    "*.{ts,tsx}": ["@parcel/validator-typescript"]
  }
}
```
and my `tsconfig.json`:
```json
{
  "compileOnSave": false,
  "compilerOptions": {
    "sourceMap": true,
    "allowSyntheticDefaultImports": true,
    "declaration": false,
    "jsx": "react",
    "esModuleInterop": true,
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "strict": true,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "target": "es5",
    "lib": [
      "es2018",
      "dom"
    ],
    "baseUrl": "./src",
    "paths": {
      "~/*": [
        "./*"
      ],
      "@cabasvert/data": [
        "../../data/src"
      ]
    }
  },
  "include": [
    "scripts",
    "src"
  ],
  "exclude": [
    "dist",
    "node_modules"
  ]
}
```

And I get:
```
@parcel/resolver-default: Cannot find module '~/file' from '/home/didier/Code/parcel2-ts-test/src'
```

**Comment by Cristy94** [comment]
I tried the `@parcel/validator-typescript` but I get this error:

```
ℹ Server running at http://localhost:1234
√ Built in 8.19s
Installing @parcel/validator-typescript...
@parcel/package-manager: Added 1 packages via npm
TypeError: input.replace is not a function
    at markdownParser (\node_modules\@parcel\markdown-ansi\lib\markdown-ansi.js:20:17)
    at prettyDiagnostic (\node_modules\@parcel\utils\lib\prettyDiagnostic.js:32:167)
    at writeDiagnostic (\node_modules\@parcel\reporter-cli\lib\CLIReporter.js:142:37)
    at _report (\node_modules\@parcel\reporter-cli\lib\CLIReporter.js:117:13)
    at Object.report (\node_modules\@parcel\reporter-cli\lib\CLIReporter.js:172:5)
    at ReporterRunner.report (\node_modules\@parcel\core\lib\ReporterRunner.js:77:31)
    at process._tickCallback (internal/process/next_tick.js:68:7)
```

**Comment by uglycoyote** [comment]
I finally got around to actually trying Parcel 2 with Typescript and I'm finding that I just get a bunch of errors from the Babel transpilation that I did not used to get from Parcel 1, and it's code that is clean as far as tsc is concerned.  (things along the lines of `@parcel/transformer-babel: myfile.tsx: Missing class properties transform.` on a fairly ordinary class member variable declaration).  Initial googling on this appeared to indicate that I needed an additional plugin for babel to solve this, but installing it did not change the situation.  I'll probably file a separate issue about this later if I cannot figure out the solution.  

I thought I would mention this here though in the context of the conversation of this discussion about whether Parcel should be running babel or tsc by default on Typescript files, since it appears to be an example of where babel is handling my files quite differently than tsc, for some reason that i don't understand, and I feel that if this happens to a lot of Typescript users it will probably turn them off from wanting to use Parcel.

**Comment by mischnic** [comment]
> Initial googling on this appeared to indicate that I needed an additional plugin for babel to solve this, but installing it did not change the situation. I'll probably file a separate issue about this later if I cannot figure out the solution.

This is because `@babel/preset-typescript` does not include the `@babel/plugin-proposal-class-properties` plugin (I'm don't know why they made that decision). You have to add it to your babelrc manually, installing is not enough

**Comment by mischnic** [comment]
> This is because `@babel/preset-typescript` does not include the `@babel/plugin-proposal-class-properties` plugin

Actually, this is because `@babel/preset-env`'s `shippedProposals` doesn't include the syntax.

**Comment by shunia** [comment]
So first, I came because I was looking for solution for the `Missing class properties transform.` error, as I just upgrade parcel@next. And thanks to @mischnic pointed that out it's a Babel problem. Big thanks dude.

And by the way, with `parcel 2.0.0-alpha.3.2`:

> ```json
> {
>   "extends": "@parcel/config-default",
>   "validators": {
>     "*.{ts,tsx}": ["@parcel/validator-typescript"]
>   }
> }
> ```

This works great with `parcel index.html`, but throws when `parcel build index.html` with:
```
Missing class properties transform.
```

> 
> ```json
> {
>   "extends": "@parcel/config-default",
>   "transformers": {
>     "*.{ts,tsx}": ["@parcel/transformer-typescript-tsc"]
>   }
> }
> ```

This throws whenever `parcel index.html` or `parcel build index.html` with:
```
Unexpected token export
/Users/xxx/Documents/xxx/node_modules/@parcel/ts-utils/src/index.js:2
export * from './FSHost';
```

**Comment by shunia** [comment]
I'm a long time Typescript user and a long time Parcel user.

It's confusing that I have to learn how to work with babel for using Parcel 2. Typescript is  compiling and type checking everything just fine, I can not find any reason to use babel.

I think it should be like: 
If people want to use babel to compile their Typescript projects, use `.parcelrc` to do so. 
If people don't know any of that, they should be just fine with their current setup (with Typescript to do the work).

Which means, use Typescript to compile should be default.

**Comment by rinick** [comment]
I'm also a long time parcel 1 user that works mostly with typescript. And this is my experience:

I tried to upgrade one  of my projects to parcel 2, spent 30 minutes, couldn't build it correctly, reverted back to 1.x.
after 6 months, did this again to see if newer version is easier, spent 30 minutes. still couldn't get it to work.

The reason I choose to switch back to parcel 1 from parcel 2 again is exactly the reason I chose parcel over webpack in the first place. I don't want to spend a lot of time learning how to config the bundler.

**Comment by DeMoorJasper** [comment]
@rinick I'm not sure I understand there's still no need to configure Parcel, the only difference with TypeScript is that we switched from tsc to Babel for performance and stability reasons.

You can always open up an issue or discussion if you think there's a bug or have a question on how anything works in Parcel 2 compared to Parcel 1. There's also a migration page on the docs site https://v2.parceljs.org/getting-started/migration/

**Comment by rinick** [comment]
@DeMoorJasper Thanks for the reply, I am finally able to run my project with parcel2. I had to delete my original .babelrc then clear parcel-cache to make it work:
`{"presets":[["@babel/env",{"targets":{"browsers":["last 1 Chrome version","last 1 firefox version"]}}]]}`

**Comment by mattrossman** [comment]
@DeMoorJasper What I'm hearing is that to use type checking with parcel you need to either:

* configure your IDE (e.g. VSCode only checks open files by default)
* configure a CI/CD pipeline
* configure a .parcerc and .tsconfig

I don't believe any of these options are "zero config".

I agree with @uglycoyote that if parcel wants to continue with this unintuitive approach, it should at least be verbose about it.

**Comment by DeMoorJasper** [comment]
@mattrossman you'll have to create a tsconfig file regardless of what you/we do, we can output a default preset but that seems like a bad idea. Some users don't use JSX, some want strict typescript, some don't, ...

Creating a file with a single line doesn't really seem like that much work to turn it on. We're still a zero-config bundler, which means you don't need config for most codebases, not that there is no config.

We're not an end-to-end development platform, we're a bundler, we could add linting and typechecking if we want and we kind of already did but it's opt-in as a lot of people still want a bundler to just bundle and not lint or typecheck

**Comment by mattrossman** [comment]
> a lot of people still want a bundler to just bundle and not lint or typecheck

@DeMoorJasper Can you create a poll to back up this claim? Because looking at the emoji reactions to the comments in this issue, it seems like most people actually do want it to typecheck. I don't think including a default tsconfig is a bad idea.

**Comment by DeMoorJasper** [comment]
> > a lot of people still want a bundler to just bundle and not lint or typecheck
> 
> 
> 
> @DeMoorJasper Can you create a poll to back up this claim? Because looking at the emoji reactions to the comments in this issue, it seems like most people actually do want it to typecheck. I don't think including a default tsconfig is a bad idea.

I guess maybe @devongovett can do this if he wants as he owns the parcel Twitter account and a lot of followers.

**Comment by mattrossman** [comment]
> Parcel is not an end-to-end development platform, it's a bundler

I should also point out that @devongovett's statements don't reflect this view.

From a [tweet in May](https://twitter.com/devongovett/status/1257728924164820992):

> I'm going to stop describing Parcel as a "bundler". That's only one small piece of what it does. In Parcel 2, it's literally a plugin!
> 
> Parcel is an application graph. It can be a compiler, a static site generator, a documentation generator, a type checker, a linter, and more.

I guess the key word here is "can" be, it just comes down to what it "should" be by default. Hopefully he can set that poll up to let the community decide.

**Comment by phaux** [comment]
> What I'm hearing is that to use type checking with parcel you need to either:
> 
>   * configure your IDE (e.g. VSCode only checks open files by default)
>   * configure a CI/CD pipeline
>   * configure a .parcerc and .tsconfig
> 
> I don't believe any of these options are "zero config".

You can simply run `tsc && parcel` instead of just `parcel` to get typechecking. It is zero config.

**Comment by mattrossman** [comment]
@phaux Not that simple:

* Right off the bat, running `tsc` does nothing in the typical Typescript folder structure. This command assumes you have configured a .tsconfig file, otherwise you need to pass additional parameters on the command line.
* This requires you to point to the compiled `.js` file in your HTML, rather than the source `.ts` entrypoint as is usually taught in the parcel guides, which adds in ambiguity (you need to know where that file will be compiled) and could throw off beginners.
* More importantly, this doesn't help during development where live reloads are important. I would expect a solution to work with `parcel index.html` and automatically re-run type checking when my files change.

**Comment by DeMoorJasper** [comment]
> This requires you to point to the compiled .js file in your HTML, rather than the source .ts entrypoint as is usually taught in the parcel guides, which adds in ambiguity (you need to know where that file will be compiled) and could throw off beginners.

I don't think there's any reason to make it this complicated, tsc will pick up the entrypoint regardless and Parcel will still do it's thing. I've done this on many projects and it's far from this complicated.

**Comment by phaux** [comment]
@mattrossman 

>   * This command assumes you have configured a .tsconfig file

Well, duh. This is how TypeScript works. Are you proposing that Parcel should ship it's own opinionated default tsconfig? Who decides which strict options are enabled, which DOM/ES libs you use, which target you want, where are your custom typings, etc?

>   * This requires you to point to the compiled `.js` file in your HTML, rather than the source `.ts` entrypoint as is usually taught in the parcel guides

What? I think I don't understand. It's enough to just run `tsc` in the console. There's a field `"include"` in tsconfig but it's optional.


>   * this doesn't help during development where live reloads are important. I would expect a solution to work with `parcel index.html` and automatically re-run type checking when my files change.

I'd say that being able to live reload an application and use it even when types are wrong is a good thing for quick prototyping. You see the errors in the IDE anyways and you can always run tsc from console or configure Parcel plugin if you want the whole project checked.

**Comment by bbugh** [comment]
I am honestly very confused about the parcel maintainers strong resistance to improving the language support! 

I definitely understand where you're coming from about Babel with fast compilation. That is compelling and I think you're right that it's a good default. What I'm less clear on and haven't seen much discussion about is why there's no typing support **at all** by default. Parcel's main agenda seems to be bring "convention over configuration" to JavaScript land... except for TypeScript? If the instructions are always going to be "install typescript, run tsc --watch" why isn't that the default zero-config setting? Why do we need yet another terminal window for yet another command when it can be run in the bundler quite easily? 😳

With a couple of changes, Parcel could offer strong TypeScript support without losing the value that your implementation Babel brings. When a `.ts` file is ran, Parcel only needs to:

- install the `typescript` package and generated a default `tsconfig.json` with `tsc --init`
- run type checking by default:
    - show non-blocking type checking during development `serve` (as a linter? with `tsc --watch --noEmit`?)
    - fail production `build` commands when there are type errors 

You still get the fast compilation with Babel and the fast development turnaround, but it operates more closely with the expectations of what "supporting typescript" really means. That's a significant improvement over what is happening now. 

What do you think?

> Well, duh. This is how TypeScript works. Are you proposing that Parcel should ship it's own opinionated default tsconfig? Who decides which strict options are enabled, which DOM/ES libs you use, which target you want, where are your custom typings, etc?

No, Parcel doesn't need to include an opinionated `tsconfig.json` because you can already generate the Microsoft opinionated default tsconfig `tsc --init` and it works fine with Parcel.

**Comment by devongovett** [comment]
I think I mostly feel like we cannot win here. Some users want Parcel to do type checking and some don’t. Parcel has never done type checking by default. Even in v1 you needed to install a plugin. I’m happy to enable it by default, but it just feels like we’ll get a bunch of issues asking us to turn it off then, especially if it isn’t a better experience than using TSC directly. 🤷🏻‍♂️ 

Regardless of the default, the TS validator plugin and the validation infrastructure in general is a bit under used atm and needs some serious work. We need to make sure it updates properly when files change, handles caching correctly, supports various TS config options, etc. Would someone on this thread be willing to help out with that? 

We’d love to make using TS with parcel a great experience but I’m not sure anyone on the core team has the time or knowledge about how to make that happen at the moment so we would very much appreciate some help. 😍

**Comment by astegmaier** [comment]
@devongovett I know I've been AWOL for the last month, but improving the type-checker is still a long-term goal of mine. FWIW, I'm also in the "turn it on by default" camp, esp. once we iron out all the kinks. I'd love some help, if anyone on this thread is interested.

**Comment by mikeplus64** [comment]
`babel-transformer` also does not seem to understand the `as` keyword in, like `foo as Type`. It is not accurate to call this TypeScript support; Parcel 2 is supporting whatever subset of it that `babel-transformer` happens to understand. I think this has the potential to be extremely frustrating to end users who expect TypeScript support but run into `babel-transformer` quirks; weighed up against that, requiring a `tsconfig.json` is really not so bad, especially since reasonable defaults can be generated straight from `tsc`. I mean, it's one of the first things that the TypeScript documentation states:

>The presence of a tsconfig.json file in a directory indicates that the directory is the root of a TypeScript project. The tsconfig.json file specifies the root files and the compiler options required to compile the project

I may simply be mistaken about Parcel's design goals, but I took "zero-configuration" to mean not that there was literally zero configuration at all, but that you wouldn't have to configure Parcel itself. As in, if I wanted Parcel to bundle a Rust project I would expect to have to have a `Cargo.toml` somewhere; that's what a Rust project is. It's a pretty common pattern on the Parcel v1 docs for various languages for them to recommend or require making configuration specific to that language; like a `.sassrc` or `.babelrc` or `bsconfig.json` etc.. Likewise, if I am coding a TypeScript project, I expect to have a `tsconfig.json`--that's what a TypeScript project is.

Just my 2c. Strongly in favour of making `tsc` the default, as `babel-transformer` doesn't (and can't) actually support TypeScript, in its entirety.

I think whether `tsc` should yell at you with type errors when `parcel` invokes it is a separate question to whether `tsc` should be used for compilation.  (Of course the answer is yes it should =), but it doesn't really matter to me since using my editor itself will yell at me for type errors as I go, and I suspect that would be the case for most people using TypeScript.) I thought the Parcel v1 pipeline of tsc->babel worked really well.

**Comment by 101arrowz** [comment]
In any case, not validating types kind of defeats the point of TypeScript. To drive the point home: it's as if you didn't throw an error when a dependency wasn't found, and instead silently replace that dependency with `null` in the generated JS; it isn't a syntactic error but it absolutely should be validated. So here's my take:

If a `tsconfig.json` is present, compile with `tsc` by default. If not, use the standard JS pipeline but validate the types after each build. My reasoning is that those who don't even spend the time to create a `tsconfig.json` are just hacking together a prototype and will prefer faster builds over safer builds. I'd be willing to create a PR for this kind of system.

**Comment by orta** [comment]
_Personally_, I don't think a bundler should run the type checker - a bundler's responsibility is to remove the types from the TS and convert it to JS as fast as possible. Putting the TS type-checker in the way of that slows the critical path a lot and adds a bunch of extra complexity to parcel when you want to make it fast (result caching, incremental builds etc )

It's likely anyone using TypeScript would already have their editors set up to show them the errors inline, and running `tsc` from the command line is a perfectly normal final check for the whole project.

**Comment by bbugh** [comment]
Also running `tsc` from the command line does not support `.vue` files which has been repeatedly mentioned here. That is not a workable solution. 

The only way to type check Vue is to build it into the build process in some way. Given that Vue 3 is in beta-almost-RC and it was rewritten in TypeScript and now has much better TS support, it's pretty important.

**Comment by argv-minus-one** [comment]
@orta I'm going to have to *completely* disagree with you. A build that is a bit sluggish but has correct results is merely inconvenient. A build that is fast but has incorrect results is useless. If I wanted “as fast as possible” and didn't care if my code is correct, I wouldn't be using TypeScript.

>It's likely anyone using TypeScript would already have their editors set up to show them the errors inline

Editors cannot show all TypeScript errors. Editors also cannot show errors that occur in files other than the one being edited.

>running `tsc` from the command line is a perfectly normal final check for the whole project.

No, it's not. Running your *build system* from the command line is a perfectly normal final check for the whole project. In a project using Parcel, that build system is Parcel, not `tsc`.

**Comment by orta** [comment]
I get that I'm posting in a thread that can realistically only be made up of people who disagree, but TypeScript always emits JS on purpose and has a compiler flag which you can turn on to get the workflow you're talking about ( [`noEmitOnError`](https://www.staging-typescript.org/tsconfig#noEmitOnError)) however the default (and how it works with tsc) is to not block JS emit on TypeScript errors.

**Comment by SamHH** [comment]
Unless I'm misunderstanding I don't believe that's relevant, you'll still get type-checking. I personally don't care if it emits or not upon type-checking failing, only that there is type-checking and any errors are reported to me.

**Comment by devongovett** [comment]
Based on the number of people on this thread, it seems supporting TS better is something we should work on. If I were to put together a plan for how we could accomplish the goals of this thread, would anyone be willing to help us implement it? There’s a lot more to this than just a config change, eg .vue/.svelte TS checking support is a big project on its own and not something that’s currently supported even with the existing plugins.

**Comment by 101arrowz** [comment]
I'd be willing to help improve TypeScript support, especially with regards to Vue. By the way, unlike in Parcel 1, TypeScript files are now actually imported as separate assets from `.vue`, which means that type-checking should be possible if the TS transformer supports it. That change was made with the [Parcel 2 Vue transformer PR](https://github.com/parcel-bundler/parcel/pull/4892).

**Comment by DeMoorJasper** [comment]
@101arrowz typechecking does not and should not happen in a transformer, it will significantly slow down building. It happens in the typescript validator https://github.com/parcel-bundler/parcel/tree/v2/packages/validators/typescript

There's a lot more to typechecking in a tool like this than just calling tsc.

**Comment by 101arrowz** [comment]
I'm not happy with performance being put ahead of safety, but since one of Parcel's core principles is being "blazing-fast," I suppose there really isn't any other option. My point is that TypeScript users, by nature of using TS over JS, have agreed to give up the rapid prototyping and monkeypatching capabilities of pure JS for more safety, and Parcel treating TS the same as JS might help prototyping but will make development more frustrating later on.

Regardless of the position the Parcel team chooses to take, the existing TS infrastructure could be improved upon, as you have mentioned, so I'll work on that for now.

**Comment by devongovett** [comment]
The reason for not doing type checking in the transformer isn’t performance, but because it’s actually impossible to do it there. Type checking requires a full graph of files. It cannot be done just on a single TS file at a time. For example, you could have imported and exported types and values. Transformer plugins in Parcel operate on individual files, as they are discovered. Type checking must be done in a Validator plugin, at the end, once the full graph is known.

**Comment by devongovett** [comment]
In order to support .vue type checking, we would need to store the TS asset that’s extracted by the transformer and somehow mark that for validation later. In the TS validator, when a .vue file is imported, we’d need to resolve that to just the TS part so that the TS compiler doesn’t need to worry about the other parts.

**Comment by devongovett** [comment]
This thread is getting very long and there's lots of different features being requested here, so I've put together three separate issues to discuss them:

1. #4936 - Implement TypeScript resolver plugin. This covers the TS specific resolution requests for `baseUrl`, `paths`, and `rootDirs`.
2. #4938 - Switch to TSC transformer instead of Babel by default. This is a simple one, but I'd like someone with a reasonably large TS project to do some perf measurements to ensure we are making the right decision. The alternative, as @bbugh said in the original issue, is to match the TSC config using Babel. This would be a short term solution that should help with some of the issues on this thread, e.g. class properties, decorators, and other features TS enables but Babel does not.
3. #4937 - An RFC for service plugins. This would allow the TypeScript compiler to be run as a service in a dedicated worker that other plugins could make requests to. This is required to fully support all TS features, including things like `const enum`s which require type information to compile and aren't supported by TSC's `transpileModule` or Babel. In addition, type errors for `.vue`/`.svelte` files could be implemented this way. This is a much longer term solution that we can begin working toward, and once it's ready, it could potentially become the default.

Please leave your feedback for the relevant feature on these individual issues. If anyone is interested in helping to implement them, please let us know. It would be greatly appreciated. 😄

**Comment by TroySchmidt** [comment]
I am happy to see there are issues now to improve the TypeScript support.  The lack of TypeScript full adoption in Parcel has been a barrier to entry for us.  
I have seen time and again people being either JavaScript or TypeScript people and neither understands the other side. If you truly use TypeScript you use it for the IDE intellisense and linting.  But you also rely on the build failing when there are typing issues as those lead to edge and corner case bugs. 
I also totally understand the decision to use Babel for transpiling and performance.  I also think a fourth issue to add is rich documentation on how to add TypeScript support using Parcel v2 and the options you have like using the validator and what the functionality and purpose that supports.  There is also adding ESLint functionality that fully supports TypeScript as well.

**Comment by kuba-orlik** [comment]
For me lack of error checking on the Parcel side makes Parcel useless to use during development. The perfect use case is that Parcel watches the files, and rebundles them when changes are detected. Errors are shown in the terminal output if any are found.

Without that functionality I have to run Typescript checker and Parcel bundler at the same time, and my PC begins to struggle

**Comment by TroySchmidt** [comment]
@kuba-orlik So the initial start of parcel `--open` with the validator setup would error on a TypeScript problem.  Also having ESLint properly setup with your IDE of choice (VS Code here) it highlights the issue.  When building the validator kicks out an error and fails the build.  The **only** time the build isn't errored out is on hot reloading.  So, @devongovett would adding the validator to the hot reloading be part of #4937 or something that could be accomplished sooner?  I also imagine it needs to take into account all validators that could error otherwise.
Considering how a setup of validator + ESLint works that is enough type checking and build failing for me to consider Parcel2 a comparable bundler to the CRA \ webpack setup.  The tooling could improve and the issues above I think are seeking to close that gap.

**Comment by kuba-orlik** [comment]
For my particular setup it would be enough if:

* typescript errors were shown in `parcel watch` output;
* typescript errors would fail the `build` command

**Comment by paeolo** [comment]
> For contrast, I've been very happy with Next.js's out-of-the-box Typescript support. Just rename a file to .tsx, and you are basically good to go. Next automatically adds a tsconfig.json if it doesn't exist.
> 
> I'm not sure if it uses tsc or Babel — either way it's working very well.

Well, I came to this thread with I think, the very same great experience in mind: Typescript support in Next.js is good. So I was convinced there was type-checking in Next.js at every-level, as my day-to-day experience with it suggests.

I was wrong.

Next.js uses webpack with babel-plugin-transform-typescript and its [caveats](https://babeljs.io/docs/en/babel-plugin-transform-typescript#caveats). So when you run Next.js in developpement mode, where things goes only through webpack, there is no type-checking and you can verify for yourself with:

`console.log(props.DOES_NOT_EXISTS)`;

On the other hand if you look at where the build code flows, you will find [verifyTypeScriptSetup](url) with this code:

`// Verify the project passes type-checking before we go to webpack phase:`
`return await runTypeCheck(ts, dir, tsConfigPath)`

> Some users want Parcel to do type checking and some don’t.

At the beginning I was in favor of it because of my experience with Next.js. But it's clear that Next.js, a kind of "zero-config" project, doesn't have it when in developpement mode (and it doesn't bother me) but have it when it comes to the build (and I think it's great).

So at the end of the day `runTypeCheck (or whatever) && parcel build` does the trick for me.

> No, it's not. Running your build system from the command line is a perfectly normal final check for the whole project. In a project using Parcel, that build system is Parcel, not tsc.

I also agree with that, oh men.

**Comment by incerta** [comment]
If someone just simply seeks a quick `TypeScript + Parcel + Jest` (with allowed `.js` files import) solution example:

```
npm i -D jest typescript ts-jest @types/jest babel-jest @babel/core @babel/preset-env
```

Add `.babelrc.js`
```js
module.exports = {
  presets: ['@babel/preset-env'],
}
```

Add `jest.config.js`
```js
module.exports = {
  transform: {
    '^.+\\.(tsx|ts)?$': 'ts-jest',
    '^.+\\.[t|j]sx?$': 'babel-jest',
  },
}
```

Add `tsconfig.json`
```js
{
  {
    "compilerOptions": {
      "module": "commonjs",
      "allowJs": true,
      "strict": true,
      "noEmit": true,
    },
    "include": ["src/**/*.ts"]
  }
}
```

From `package.json`
```js
...
  "scripts": {
    "start": "parcel serve ./src/index.html --port 3000",
    "build": "tsc --noEmit && jest && parcel build ./src/index.html",
    "type-check": "tsc --noEmit --watch",
    "test": "jest --watch"
  },
...
```

**Comment by LukeTOBrien** [comment]
Hello there,

I think you've goon off topic abit, let me bring it back.

The point I make in my issue:  

> **What % of TypeScript projects use a tsconfig.json?**  
I use a tsconfig in 100% of my TypeScript project, so in my case at least I am always going to manually configure to use tsc (which I never did in Parcel 1, it just worked)

Also:

> for most cases developers will want to **transform TypeScript as nature intended using tsc**.

I think Parcel 2 should work the same as Parcel 1 with regard to TypeScript.  
The reason that Babel is used is that it is a bit faster, but I am not concerned with speed, Parcel is fast enough, and TBH even if it takes seconds or minutes then I don't mind.

**Comment by LukeTOBrien** [comment]
@mischnic I do like some of the quotes above:

* A build that is fast but has incorrect results is **useless**.
* For me lack of error checking on the Parcel side makes Parcel **useless** to use during development.

I do hope the developers listen and will switch to `tsc` by default.

**Comment by mischnic** [comment]
> I think Parcel 2 should work the same as Parcel 1 with regard to TypeScript.

Parcel 1 didn't do typechecking either.

**Comment by LukeTOBrien** [comment]
I did notice that, but Parcel 1 used the `tsconfig`.  
How does Parcel 2 handle React?... I do not use Reat myself but the [Parcel 1 docs](https://parceljs.org/typeScript.html#when-using-react) show you need a option in tsconfig.  
Presumably using React in Parcel 2 would require the `tsc` config in the `.rc`?

```javascript
// tsconfig.json
{
  "compilerOptions": {
    "jsx": "react"
  }
}
```

**Comment by mischnic** [comment]
The default Babel config generated by Parcel 2 would include `@babel/preset-env`, `@babel/preset-typescript`, `@babel/preset-react` in that case.

**Comment by LukeTOBrien** [comment]
I have created the `.parcelrc` config required for using TypeScript... I had two further errors because Parcel couldn't find the modules `@parcel/transformer-typescript-tsc` and `typescript`, so I had to `npm i` thouse.  
I would say this is a breaking change from anyone who used Parcel 1 and it breaks the 'no congfiguration' pholisophy.

**Comment by devongovett** [comment]
You are welcome to submit a PR. Further complaining about this issue is pointless. We've created the issues linked in my above comment to make our intent clear. If you would like to help make that happen faster, please do.

**Comment by devongovett** [comment]
At this point, I'm going to close this. Parcel is not a type checker and will probably never be one. I don't have the bandwidth to implement and maintain this but if someone wants to implement it as a third party plugin that would be great (I guess it could probably be implemented as a reporter). Otherwise, I recommend simply running `tsc --noEmit` during your production build script and using TypeScript in your IDE during development.
